p2p/tools/local_network.py

```diff
@@ -0,0 +1,378 @@
+import asyncio
+import os
+import random
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    Iterable,
+    NamedTuple,
+    List,
+    Tuple,
+)
+
+import pytest
+
+from p2p.cancel_token import CancelToken, wait_with_token
+from p2p.exceptions import OperationCancelled
+from p2p.service import BaseService
+
+
+class Address(NamedTuple):
+    host: str
+    port: int
+
+    def __repr__(self):
+        return '<%s %s>' % (self.__class__.__name__, self)
+
+    def __str__(self):
+        return '%s:%s' % (self.host, self.port)
+
+
+class MemoryTransport(asyncio.Transport):
+    """
+    Direct connection between a StreamWriter and StreamReader.
+    """
+    _reader: asyncio.StreamReader
+
+    def __init__(self, reader: asyncio.StreamReader) -> None:
+        super().__init__()
+        self._reader = reader
+
+    def write(self, data: bytes) -> None:
+        self._reader.feed_data(data)
+
+    def writelines(self, data: Iterable[bytes]) -> None:
+        for line in data:
+            self.write(line)
+            self.write(b'\n')
+
+    def write_eof(self) -> None:
+        self._reader.feed_eof()
+
+    def can_write_eof(self) -> bool:
+        return True
+
+    def is_closing(self) -> bool:
+        return False
+
+    def close(self) -> None:
+        self.write_eof()
+
+
+class AddressedTransport(MemoryTransport):
+    """
+    Direct connection between a StreamWriter and StreamReader.
+    """
+    _queue: asyncio.Queue
+
+    def __init__(self, address: Address, reader: asyncio.StreamReader) -> None:
+        super().__init__(reader)
+        self._address = address
+        self._queue = asyncio.Queue()
+
+    @property
+    def queue(self) -> asyncio.Queue:
+        return self._queue
+
+    def get_extra_info(self, name: str, default=None):
+        if name == 'peername':
+            return (self._address.host, self._address.port)
+        else:
+            return super().get_extra_info(name, default)
+
+    def write(self, data: bytes) -> None:
+        super().write(data)
+        self._queue.put_nowait(len(data))
+
+
+ReaderWriterPair = Tuple[asyncio.StreamReader, asyncio.StreamWriter]
+
+
+def addressed_pipe(address: Address) -> ReaderWriterPair:
+    reader = asyncio.StreamReader()
+
+    transport = AddressedTransport(address, reader)
+    protocol = asyncio.StreamReaderProtocol(reader)
+
+    writer = asyncio.StreamWriter(
+        transport=transport,
+        protocol=protocol,
+        reader=reader,
+        loop=asyncio.get_event_loop(),
+    )
+    return reader, writer
+
+
+def direct_pipe() -> ReaderWriterPair:
+    reader = asyncio.StreamReader()
+
+    transport = MemoryTransport(reader)
+    protocol = asyncio.StreamReaderProtocol(reader)
+
+    writer = asyncio.StreamWriter(
+        transport=transport,
+        protocol=protocol,
+        reader=reader,
+        loop=asyncio.get_event_loop(),
```

Comment:
Probably want the ability to override the default loop with an optional kwarg to `direct_pipe`.